solr/contrib/ltr/src/test/org/apache/solr/ltr/TestCacheInteractionOfPrefetchingFieldValueFeature.java

```diff
@@ -0,0 +1,530 @@
+    // needed to clear cache because we make assertions on its content
+    reloadCollection(COLLECTION);
+
+    SolrQuery query = new SolrQuery("{!func}sub(8,field(popularity))");
+    query.setRequestHandler("/query");
+    query.setParam("rows", "8");
+    query.setFields("id,features:[fv]");
+    query.add("rq", "{!ltr model=powpularityS-model reRankDocs=8}");
+
+    QueryResponse queryResponse = solrCluster.getSolrClient().query(COLLECTION,query);
+
+    Map<String, List<List<String>>> loadedFields = ObservingPrefetchingFieldValueFeature.loadedFields;
+
+    for (SolrDocument doc : queryResponse.getResults()) {
+      String docId = (String) doc.getFirstValue("id");
+      if (docId.equals("1")) {
+        assertEquals(NUM_FEATURES, loadedFields.get(docId).stream()
+            // doc with id 1 has no values set for 3 of the 6 feature fields
+            .filter(fieldLoadedList -> fieldLoadedList.size() == NUM_FEATURES - 3)
+            .count());
+      } else {
+        // all the fields were loaded at once
+        assertEquals(NUM_FEATURES, loadedFields.get(docId).stream()
+            .filter(fieldLoadedList -> fieldLoadedList.size() == NUM_FEATURES)
+            .count());
+      }
+    }
+    assertTheResponse(queryResponse);
+  }
+
+  @Test
+  public void testSimpleQueryBreakPrefetching() throws Exception {
+    ObservingPrefetchingFieldValueFeature.setBreakPrefetching(true);
+    ObservingPrefetchingFieldValueFeature.loadedFields = new HashMap<>();
+    System.setProperty(LAZY_FIELD_LOADING_CONFIG_KEY, "false");
+    // needed to clear cache because we make assertions on its content
+    reloadCollection(COLLECTION);
+
+    SolrQuery query = new SolrQuery("{!func}sub(8,field(popularity))");
+    query.setRequestHandler("/query");
+    query.setParam("rows", "8");
+    query.setFields("id,features:[fv]");
+    query.add("rq", "{!ltr model=powpularityS-model reRankDocs=8}");
+
+    QueryResponse queryResponse =
+        solrCluster.getSolrClient().query(COLLECTION,query);
+
+    Map<String, List<List<String>>> loadedFields = ObservingPrefetchingFieldValueFeature.loadedFields;
+
+    assertEquals(loadedFields.size(), queryResponse.getResults().size());
+    for (SolrDocument doc : queryResponse.getResults()) {
+      String docId = (String) doc.getFirstValue("id");
+      if (docId.equals("1")) {
+        // doc with id 1 has no values set for 3 of the 6 feature fields
+        assertEquals(NUM_FEATURES - 3, loadedFields.get(docId).stream()
+            .filter(fieldLoadedList -> fieldLoadedList.size() == 1)
+            .count());
+      } else {
+        // each single field used for a feature gets loaded separately
+        assertEquals(NUM_FEATURES, loadedFields.get(docId).stream()
+            .filter(fieldLoadedList -> fieldLoadedList.size() == 1)
+            .count());
+      }
+    }
+    assertTheResponse(queryResponse);
+  }
+
+  @Test
+  public void testSimpleQueryLazyBreakPrefetching() throws Exception {
```

Comment:
At a glance it's tricky to see how much overlap there is between the tests in this class here but conceptually if two tests `testOdd` and `testEven` are almost identical it could be helpful to factor out a common implementation method to reduce code duplication e.g.

```
public void testOdd() {
  implTestOddOrEven(true);
}

public void testEven() {
  implTestOddOrEven(false);
}

private void implTestOddOrEven(boolean odd) {
  ...
}
```

Reply:
I hope that the commit to make the tests more comprehensible ( 2d355c7 ) helps with identifying overlap.
The `BreakPrefetching` and `DisablePrefetching` tests are still very similar. They are not intended to test different behavior but rather that different configurations lead to the same behavior.

Reply:
I extracted some more code and added explanations for the thoughts behind the different tests ( 8d0ab32 ).